<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">



<title>GWAS workflows with MTMC-SKAT</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">GWAS workflows with MTMC-SKAT</h1>



<p>In this vignette, we describe key mid-level functions provided by the <code>MTMC-SKAT</code> package, their usage in the standard automated workflow and how they can be used to produce a custom workflow.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="co">#library(SKATMCMT)</span></span></code></pre></div>
<p>The function <code>pre_allocate</code> is used to load SNP data from the <code>.traw</code> format and convert it to a list of SNP windows. The resulting list is saved in the directory specified by <code>pre_allocated_dir</code>. To avoid regenerating a list used for multiple analyses, <code>pre_allocate</code> will read in the corresponding list for a genotype file if one exists.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a>pre_allocated_SNP_windows &lt;-<span class="st"> </span><span class="kw">pre_allocate</span>(<span class="dt">raw_file_path =</span> <span class="st">&quot;poplar_Chr10_portion.traw&quot;</span>,</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a>                                          <span class="dt">window_size =</span> <span class="dv">3000</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a>                                          <span class="dt">window_shift =</span> <span class="dv">1000</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a>                                          <span class="dt">pre_allocated_dir =</span> <span class="st">&quot;pre_allocated/&quot;</span>)</span></code></pre></div>
<p>Upon pre-allocating SNP window data, we can begin the initial round of SKAT, without resampling, to produce inital p-values for each SNP window without any permutations. From these initial results, the user or preprogrammed filters can select SNP windows deemed interesting and perform <code>MTMC-SKAT</code> to calculate empirical p-values for these SNP windows. The provided function for multi-threaded SKAT (<code>mtskat</code>) runs SKAT over pre-allocated SNP windows, with chunks of SNP windows distributed among cores.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a>results_sans_resampling &lt;-<span class="st"> </span><span class="kw">mtskat</span>(<span class="dt">pre_allocated_SNP_windows =</span> pre_allocated_SNP_windows</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>                                  <span class="dt">phenotype =</span> <span class="kw">read.csv</span>(<span class="st">&quot;poplar_shoot_sample.csv&quot;</span>),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a>                                  <span class="dt">covariates =</span> <span class="kw">read.csv</span>(<span class="st">&quot;poplar_PC_covariates.csv&quot;</span>),</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a>                                  <span class="dt">ncore =</span> <span class="st">&quot;AllCores&quot;</span>)</span></code></pre></div>
<p>Based on these initial results, batches of SNP windows producing p-values over a given range of interest can be selected using <code>extract_windows_range_p</code>. As in the standard automated workflow, we recommend the range of p-values for each batch of SNPs vary by no more than an order of magnitude, to allow calculation of empirical p-values to a desired accuracy without an unnecesary number of permutations.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a>window_list &lt;-<span class="st"> </span><span class="kw">extract_windows_range_p</span>(<span class="dt">data =</span> results_sans_resampling,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a>                                       <span class="dt">upper_bound =</span> <span class="dv">10</span><span class="op">^-</span><span class="dv">2</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a>                                       <span class="dt">lower_bound =</span> <span class="dv">10</span><span class="op">^-</span><span class="dv">3</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a>new_pre_allocated_SNP_windows &lt;-<span class="st"> </span><span class="kw">subset_list_of_lists</span>(<span class="dt">list_of_lists =</span> pre_allocated_SNP_windows,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a>                                                      <span class="dt">desired_list =</span> window_list)</span></code></pre></div>
<p>Once this subset of SNP windows has been pre-allocated into a new list, empirical p-values can be calculated using <code>mtmcskat</code> with the <code>multithreading</code> option set to <code>snp</code> (to parallelize over SNP windows) or <code>nm</code> (to parallelize over permuted null models). The former is more efficient when 1) a sufficient number of permuted null models for the desired accuracy of empirical p-values can be stored in memory; and 2) The number of SNP windows is no less than the number of cores. For lower p-value ranges requiring larger number of permutations and with smaller batches of SNP windows, the latter mode becomes more efficient. In the standard automated workflow, the appropriate function is used for each batch of SNPs based on these criteria.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a>batch_results_with_resampling_over_SNPs &lt;-<span class="st"> </span><span class="kw">mtmcskat</span>(<span class="dt">pre_allocated_SNP_windows =</span> new_pre_allocated_SNP_windows,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a>                                                    <span class="dt">phenotype =</span> this_phenotype,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a>                                                    <span class="dt">covariates =</span> covariates,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a>                                                    <span class="dt">ncore =</span> <span class="st">&quot;AllCores&quot;</span>,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a>                                                    <span class="dt">multithreading =</span> <span class="st">&quot;snp&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a>batch_results_with_resampling_over_NMs &lt;-<span class="st"> </span><span class="kw">mtmcskat</span>(<span class="dt">pre_allocated_SNP_windows =</span> new_pre_allocated_SNP_windows,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a>                                                    <span class="dt">phenotype =</span> this_phenotype,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a>                                                    <span class="dt">covariates =</span> covariates,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a>                                                    <span class="dt">ncore =</span> <span class="st">&quot;AllCores&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a>                                                    <span class="dt">multithreading =</span> <span class="st">&quot;nm&quot;</span>)</span></code></pre></div>
<p>In rare cases, empirical p-values may be significantly lower than initial p-values, leading to fewer significant figures for an empirical p-value than expected given the number of permutations run. These SNP windows can then be extracted for subsequent rounds of analyses with more permutations, until the desired number of significant figures or accuracy limit is reached.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a>window_list &lt;-<span class="st"> </span><span class="kw">extract_windows_sig_fig</span>(<span class="dt">data =</span> batch_results_with_resampling_over_NMs,</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a>                                       <span class="dt">n_sig_fig =</span> <span class="dv">1</span>) <span class="co"># Suppose we desire two significant figures and thus want all windows for which empirical p-values only have one</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a>new_pre_allocated_SNP_windows &lt;-<span class="st"> </span><span class="kw">subset_list_of_lists</span>(<span class="dt">list_of_lists =</span> pre_allocated_SNP_windows,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a>                                                      <span class="dt">desired_list =</span> window_list)</span></code></pre></div>
<p>…</p>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
